### ISSUE TITLE
Struggling when passing JSON to --extra-vars

### ISSUE THREAD

**Issue by rik2803** [rundeck-plugins/ansible-plugin#26]
Hi,

I am trying to run a playbook for which I need to pass a JSON structure to `--extra-vars`. The only way I can make this work is when no space is in the argument to `--extra-vars`.

This works:

`--verbose --tags=some_tag --extra-vars='{"springbootapps":{"appname":{"version":"0.1.17","username":"someuser","urlbase":"https://some.site.com/a/path/to/a/file","nexususer":"someusername","nexuspasswd":"apassword","nexusname":"warname-0.1.17.war"}}}'`

Putting a space inside the JSON string (anywhere) results in the plugin to treat it as a new commandline options (I think).

Is there a way to escape the spaces? Allowing spaces would increase legibility, and sometimes, I would need to have spaces inside some value in the JSON string.

Thanks!

**Comment by reecewebb** [comment]
I can confirm this. Something broke between 1.2.2 and 1.2.3... With 1.2.2 I could pass in multiple extra variables to the playbook, such as:

-t ping -e "host=${option.Host} mailcf=${option.Configuration}"

Now with 1.2.3 that space between var's causes an immediate failure. The Ansible error is:

ERROR! the playbook: mailcf=Research could not be found

("Research" is the value of ${option.Configuration}).

**Comment by suivethefirst** [comment]
"-e 'var=${option.var} var2=${option.var2}'"

This is the workaround I have used since 1.2.3 (I.e., include the double and single quotes).

**Comment by frozenice** [comment]
Yeah, argument processing is a pain point ([pointer here](https://github.com/Batix/rundeck-ansible-plugin/blob/master/src/main/java/com/batix/rundeck/AnsibleRunner.java#L205)).

The problem is that we get a `String` from Rundeck for the extraArgs option, but the `ProcessBuilder` wants its arguments as a `List<String>`. So we need to correctly split the String up into pieces, which resemble the single arguments.

So `-t ping -e "some thing"` would need to become `["-t", "ping", "-e", "some thing"]`, but don't forget nested and escaped quotes... I tried using some helper classes, but I guess something is not quite right.

Needs to be looked at carefully. PRs welcome. :)

**Comment by ahonor** [comment]
@frozenice Curious if the rundeck plugin system could be enhanced to use multi-value config properties to help this matter. This idea would allow a user to add a list of values. It could work similarly to the multi-value job option.

**Comment by frozenice** [comment]
@ahonor Yes, this would be nice!

**Comment by ahonor** [comment]
@frozenice https://github.com/rundeck/rundeck/issues/1847

**Comment by frozenice** [comment]
Can you please try the new [v1.2.4](https://github.com/Batix/rundeck-ansible-plugin/releases/tag/1.2.4)? See the [README](https://github.com/Batix/rundeck-ansible-plugin#debugging) on how to enable debugging!

I changed some args parsing stuff there.

**Comment by frozenice** [comment]
Any update on this? There were newer versions released in the meantime, so test the newest version please.

**Comment by ericpfisher** [comment]
@frozenice Thanks for the follow-up ask.  I'm still struggling with passing JSON `--extra-vars` to the Ansible command via "Extra Arguments" in my job configuration.  Using version 1.3.0 of the plugin:

```
[admin@rundeck ~]$ ls /var/lib/rundeck/libext/ | grep ansible
ansible-plugin-1.3.0.jar
```

I'm attempting to pass a list as a value of an extra variable, traditionally done using JSON from the command line like so:

```
ansible-playbook --extra-vars '{"args":["arg1","arg2"]}' foo-playbook.yml
```

When I use functional Ansible command line syntax in the "Extra Arguments", I'm finding that my single quotes are removed, and my double quotes are automatically escaped now (unquoted values in the list is intentional and handled by my job options, unrelated to the issue):

```
...
14:20:19            extraArgs: -vvv --check --extra-vars='{"args":[foo,bar]}'
14:20:19            tokenized: [-vvv, --check, --extra-vars={\"args\":[foo,bar]}]
14:20:19            Ansible command:
14:20:19            [ansible-playbook, /playbooks/foo.yml, -l, foo.host.com, -vvvv, -vvv, --check, --extra-vars={\"args\":[foo,bar]}]
...
```

**NOTE**: I've tried with and without `=` joining `--extra-vars` with the variable values.  Same result.

This causes the playbook to NOT receive the values for `args` and Ansible then uses the default I have in my playbook (which is undesired...I WANT it to use the values I pass through the job configuration "Extra Arguments").

I've gotten close to the desired JSON `--extra-vars` value by escaping the single quotes that wrap the entire JSON object, but I'm guessing the [redaction of beginning/ending double quotes in in `tokenizeCommand`](https://github.com/Batix/rundeck-ansible-plugin/blob/master/src/main/java/com/batix/rundeck/AnsibleRunner.java#L53) is interfering when I want to wrap the JSON object keys in double quotes (I'm comfortable with Regex, but not familiar with Java).

Let me know if I'm doing something obviously wrong, or if more information is needed.

**Comment by yassineazzouz** [comment]
Hi @ericpfisher ,

Do you still have this issue with the latest version ?

Thanks,

**Comment by yassineazzouz** [comment]
Hi All,

This is fixed in v2.0.1

Thanks,
